Source/WebCore/platform/graphics/BitmapImageSource.cpp

```diff
@@ -0,0 +1,484 @@
+        LOG(Images, "BitmapImageSource::%s - %p - url: %s. Decoding for frame at index = %d will be requested.", __FUNCTION__, this, sourceUTF8(), index);
+        frameProvider->requestNativeImageAtIndex(index, SubsamplingLevel::Default, ImageAnimatingState::No, { DecodingMode::Asynchronous });
+        return;
+    }
+
+    LOG(Images, "BitmapImageSource::%s - %p - url: %s. Frame at index = %d was decoded for natural size.", __FUNCTION__, this, sourceUTF8(), index);
+    callDecodeCallbacks(DecodingStatus::Complete);
+}
+
+void BitmapImageSource::callDecodeCallbacks(DecodingStatus status)
+{
+    if (m_decodeCallbacks.isEmpty())
+        return;
+    for (auto& decodeCallback : m_decodeCallbacks)
+        decodeCallback(status);
+    m_decodeCallbacks.clear();
+}
+
+void BitmapImageSource::imageFrameDecodeAtIndexHasFinished(unsigned index, ImageAnimatingState animatingState, DecodingStatus decodingStatus)
+{
+    // Depending on its timer, the animator may call or postpone calling imageFrameAvailable().
+    if (m_animator && m_animator->imageFrameDecodeAtIndexHasFinished(index, animatingState, decodingStatus))
+        return;
+
+    imageFrameAtIndexAvailable(animatingState, decodingStatus);
+}
+
+void BitmapImageSource::imageFrameAtIndexAvailable(ImageAnimatingState animatingState, DecodingStatus decodingStatus)
+{
+    callDecodeCallbacks(decodingStatus);
+
+    if (decodingStatus == DecodingStatus::Invalid)
+        return;
+
+    if (auto observer = m_bitmapImage.imageObserver())
+        observer->imageFrameAvailable(m_bitmapImage, animatingState, nullptr, decodingStatus);
+}
+
+const ImageFrame& BitmapImageSource::frameAtIndexCacheIfNeeded(unsigned index)
+{
+    if (auto frameProvider = ensureFrameProvider())
+        return frameProvider->frameAtIndexCacheIfNeeded(index);
+    return ImageFrame::defaultFrame();
+}
+
+RefPtr<NativeImage> BitmapImageSource::nativeImageAtIndexCacheIfNeeded(unsigned index, SubsamplingLevel subsamplingLevel, const DecodingOptions& options)
+{
+    auto frameProvider = ensureFrameProvider();
+    if (!frameProvider)
+        return nullptr;
+
+    auto nativeImage = frameProvider->nativeImageAtIndexCacheIfNeeded(index, subsamplingLevel, options);
+    if (!nativeImage)
+        return nullptr;
+
+    return RefPtr { nativeImage->ptr() };
+}
+
+RefPtr<NativeImage> BitmapImageSource::preTransformedNativeImageAtIndex(unsigned index, ImageOrientation orientation)
+{
+    auto nativeImage = nativeImageAtIndex(index);
+    if (!nativeImage)
+        return nullptr;
+
+    auto size =  this->size();
+    auto sourceSize = this->sourceSize();
+
+    if (orientation == ImageOrientation::Orientation::None && size == sourceSize)
+        return nativeImage;
+
+    if (orientation == ImageOrientation::Orientation::FromImage)
+        orientation = frameOrientationAtIndex(index);
+
+    auto buffer = ImageBuffer::create(size, RenderingPurpose::Unspecified, 1, DestinationColorSpace::SRGB(), PixelFormat::BGRA8);
```

Comment:
Existing issue, but why do we use `SRGB()` here rather than try to use the color space of the source bitmap (if the color space can be used as a destination)?

Reply:
I think what we care about is the `ColorSpace` of the `GraphicsContext` on which this image will be drawn to. But we do not have access to it here.

Reply:
Yes, choosing the `GraphicsContext` destination color space will be correct, but won't we possibly lose some color resolution by converting to sRGB here, instead of trying to maintain the bitmap's original color space (assuming that it is possible to use it as a destination color space)?